lightning/mydump/parquet_parser.go

```diff
@@ -0,0 +1,192 @@
+	batchReadRowSize = 32
+)
+
+type ParquetParser struct {
+	Reader   *preader.ParquetReader
+	columns  []string
+	rows     []interface{}
+	readRows int64
+	curStart int64
+	curIndex int
+	lastRow  Row
+	logger   log.Logger
+}
+
+type ReaderWrapper struct {
+	ReadSeekCloser
+}
+
+func (r *ReaderWrapper) Write(p []byte) (n int, err error) {
+	return 0, errors.New("unsupported operation")
+}
+
+func (r *ReaderWrapper) Open(name string) (source.ParquetFile, error) {
+	return nil, errors.New("unsupported operation")
+}
+func (r *ReaderWrapper) Create(name string) (source.ParquetFile, error) {
+	return nil, errors.New("unsupported operation")
+}
+
+func NewParquetParser(path string) (*ParquetParser, error) {
+	r, err := local.NewLocalFileReader(path)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// FIXME: need to bench what the best value for the concurrent reader number
+	reader, err := preader.NewParquetReader(r, nil, 2)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	columns := make([]string, 0, len(reader.Footer.Schema))
+	for i, c := range reader.Footer.Schema {
+		if c.GetNumChildren() == 0 {
+			// the SchemaElement.Name is capitalized, we should use the original name
+			columns = append(columns, reader.SchemaHandler.Infos[i].ExName)
+		}
+	}
+
+	return &ParquetParser{
+		Reader:  reader,
+		columns: columns,
+		logger:  log.L(),
+	}, nil
+}
+
+// Pos returns the currently row number of the parquet file
+func (pp *ParquetParser) Pos() (pos int64, rowID int64) {
+	return pp.curStart + int64(pp.curIndex), pp.lastRow.RowID
+}
+
+func (pp *ParquetParser) SetPos(pos int64, rowID int64) {
+	if pos < pp.curStart {
+		panic("don't support seek back yet")
+	}
+	pp.lastRow.RowID = rowID
+
+	if pos < pp.curStart+int64(len(pp.rows)) {
+		pp.curIndex = int(pos - pp.curStart)
+		pp.readRows = pos
+		return
+	}
+
+	if pos > pp.curStart+int64(len(pp.rows)) {
+		pp.Reader.SkipRows(pos - pp.curStart - int64(len(pp.rows)))
+	}
+	pp.curStart = pos
+	pp.readRows = pos
+	pp.curIndex = 0
+	if len(pp.rows) > 0 {
+		pp.rows = pp.rows[:0]
+	}
+}
+
+func (pp *ParquetParser) Close() error {
+	return pp.Reader.PFile.Close()
+}
+
+func (pp *ParquetParser) ReadRow() error {
+	pp.lastRow.RowID++
+	if pp.curIndex >= len(pp.rows) {
+		if pp.readRows >= pp.Reader.GetNumRows() {
+			return io.EOF
+		}
+		count := batchReadRowSize
+		if pp.Reader.GetNumRows()-pp.readRows < int64(count) {
+			count = int(pp.Reader.GetNumRows() - pp.readRows)
+		}
+
+		var err error
+		var rows []interface{}
+		//if len(pp.rows) < count {
+		//	rows, err = pp.Reader.ReadByNumber(count)
+		//	pp.rows = rows
+		//} else {
+		//	pp.rows = pp.rows[:count]
+		//	err = pp.Reader.Read(&pp.rows)
+		//}
+		rows, err = pp.Reader.ReadByNumber(count)
+		pp.rows = rows
+		if err != nil {
+			return errors.Trace(err)
+		}
+		pp.curStart = pp.readRows
+		pp.readRows += int64(len(pp.rows))
+		pp.curIndex = 0
+	}
+
+	row := pp.rows[pp.curIndex]
+	pp.curIndex++
+
+	v := reflect.ValueOf(row)
```

Comment:
can we do a benchmark? i'm afraid that `xitongsys/parquet-go/reader` wrapping it into `reflect.Value` and we here immediately unwrapping it here would cost a lot

Reply:
Previous test result shows these reflect do have performance impact, maybe we can do some optimization later.